File: Prefix_Sums/Silver/P3_2015-Breed_Counting/P3_2015-Breed_Counting.py
# Prefix sums approach
# T & M: O(n + q), where n is num of cows, and q is num of queries

# Open input file
with open("bcount.in") as f:
	# first line: n = num of cows, q = num of queries
	n, q = map(int, f.readline().split())

	# For next n lines read cow breed IDs and create prefix sum lists
	holsteins = [0] * (n + 1)
	guernseys = [0] * (n + 1)
	jerseys = [0] * (n + 1)

	for idx in range(1, n+1):
		holsteins[idx] += holsteins[idx - 1]
		guernseys[idx] += guernseys[idx - 1]
		jerseys[idx] += jerseys[idx - 1]
		cow_id = int(f.readline().strip())
		if cow_id == 1:
			holsteins[idx] += 1
		elif cow_id == 2:
			guernseys[idx] += 1
		elif cow_id == 3:
			jerseys[idx] += 1

	# For next q lines solve num of cow breeds in query range
	queries_ids = []
	for ln in range(q):
		start, end = map(int, f.readline().split())
		queries_ids.append([holsteins[end] - holsteins[start-1], guernseys[end] - guernseys[start-1], jerseys[end] - jerseys[start-1]])
		
# Write query ids to output file
with open("bcount.out", "w") as f:
	for q_ids in queries_ids:
		f.write(" ".join(map(str, q_ids)) + "\n")

File: Arrays_and_Hashing/Prefix_Sums/Silver/P3_2015-Breed_Counting/P3_2015-Breed_Counting.cpp
// Prefix sums approach
// T: O(n + q), M: O(n), where n is num of cows, and q is num of queries

// #include <bits/stdc++.h>

#include <iostream>

using namespace std;

int main() {
	// Open input file
	freopen("bcount.in", "r", stdin);
	// first line: n = num of cows, q = num of queries
	int n, q;
	cin >> n >> q;

	// For next n lines read cow breed IDs and create prefix sum lists
	int holsteins[n+1] = {}, guernseys[n+1] = {}, jerseys[n+1] = {};
	for (int idx = 1; idx < n + 1; idx++) {
		holsteins[idx] += holsteins[idx-1];
		guernseys[idx] += guernseys[idx-1];
		jerseys[idx] += jerseys[idx-1];
		int cow_id;
		cin >> cow_id;
		switch (cow_id) {
			case 1:
				holsteins[idx]++;
				break;
			case 2:
				guernseys[idx]++;
				break;
			case 3:
				jerseys[idx]++;
				break;
		}
	}

	// For next q lines solve num of cow breeds in query range and write queries ids to output file
	freopen("bcount.out", "w", stdout);
	for (int ln = 0; ln < q; ln++) {
		int start, end;
		cin >> start >> end;
		cout << holsteins[end] - holsteins[start-1] << " "
			<< guernseys[end] - guernseys[start-1] << " "
		    << jerseys[end] - jerseys[start-1] << endl;
	}
}
